### DISCUSSION TITLE
Writing a lua script for YimMenu

### DISCUSSION THREAD

**Discussion by MindOfBog** [YimMenu/YimMenu#3602] [Q&A] [answered]
I apologize in advance if this question has been asked before, but I haven't been able to find a good resource explaining things for this. I am thinking of writing a menu for some specific niche things that none of the other scripts I've tested have. I have looked through some other scripts and the basic stuff like imGUI I am probably fine with.

What I actually need help with is:
- how do I actually create mods? 
- Do I need to find memory offsets and how do I do that? 

Is there a general guide or resource for learning that stuff?

**Comment by xesdoog** [comment, marked as answer]
> how do I actually create mods?

- The first thing you need is a problem solving brain. That comes naturally and can be improved with practice. It doesn't directly define your programming skills, it just controls your ability to make and fix things in general whether it is programming, arts, craftsmanship, etc...

- Second you have to learn Lua. If you don't know what variables and data types are then you won't be able to do anything other than copy / paste other people's work (that's frowned upon unless you ask for permission and give credits). Sure, you can learn a few tips and tricks by looking at / experimenting with other people's code but if you copy for example `globals.set_int(some_global, some_offset, some_number)` and you don't know that `some_global`, `some_offset`, `some_number` are variable integers that were defined somewhere in the script and you don't know what **[`globals.set_int()`](https://github.com/YimMenu/YimMenu/blob/master/docs/lua/tables/globals.md)** even is then you'll be stuck looking at something you don't understand which is exactly what happens to me when I look at C++ code because I don't know cpp and never got myself to sit down and learn even the basics of it. Start here to [learn Lua](https://lua.org/pil) itself (you need a basic understanding of programming in general) then read [YimMenu's docs](https://github.com/YimMenu/YimMenu/tree/master/docs/lua).

- Third you have to know exactly what you're trying to achieve. As for how to achieve it, a bit of research will get you there. Here are a few references to get you started:
  - [GTA Forums](https://gtaforums.com/)
  - [UnknownCheats](https://www.unknowncheats.me/forum/grand-theft-auto-v/)
  - [Pleb Masters](https://forge.plebmasters.de/objects)
  - [RAGEMP](https://wiki.rage.mp/index.php?title=Main_Page)
  - [cfx.re](https://forum.cfx.re/) <- This is for FiveM but a lot of things still apply to GTAO.

> Do I need to find memory offsets and how do I do that?
- You only need to do that if you want to interact with memory which I doubt it. What I think is happening here, since you mentioned looking at **Casino Pacino**, is that you're confusing the game's scripts global / local offsets with memory offsets. Read the docs to understand what is what.
- You can find globals and locals along with their offsets already dumped [here](https://www.unknowncheats.me/forum/grand-theft-auto-v/500059-globals-locals-discussion-read-page-1-a.html) or you can look through [decompiled scripts](https://github.com/root-cause/v-decompiled-scripts/) and find them yourself.
- As for memory stuff, I'm not good with that myself so I can't really help you but what I do know is that programs store values in your memory. You can read those by scanning values that you know in Cheat Engine which when found will give you the offset but the memory address itself changes with every game launch so you'll have to find the base address. Let's suppose you know that your game stores your health for example as 4 bytes, your health is 100 in-game; you can scan for a 4 byte value of 100, then decrease it and do the next scan for a decreased value, rinse and repeat until you find where your health is stored. An offset is basically how far that value is stored.
Another (more advanced) way of doing it is to dump the game's executable using a decompiler like IDA, Ghidra, Binary Ninja, etc... (they each have their pros and cons). I'm not a reverse-engineer so I can't give you any more advice or information about this but [this thread](https://www.unknowncheats.me/forum/grand-theft-auto-v/144028-grand-theft-auto-reversal-structs-offsets.html) and [this repo](https://github.com/Yimura/GTAV-Classes/tree/master) could probably help.

**OR** you can scratch all of the above and just become a skibbidy skidd. It's up to you.

**Reply by MindOfBog** [reply]
Thank you so much, these pointers are great! While I haven't really done much in lua before, I'm not new to programming and it seems decently straight forward. I have looked through some scripts and UC posts as well as looked through some of the documentation for Yim and doing basic stuff is very easy, what I'm trying to do is messing with in game menus and stuff so I will continue digging until I find what I'm looking for.

P.S. Also I'm not a skid and I've always tried to implement things myself instead of copying code 1:1, and if I do need to use someone elses work I will make sure I have permission from the author and credit them for it as well lol.

**Reply by xesdoog** [reply]
> what I'm trying to do is messing with in game menus and stuff

You mean the game's menus or YimMenu's UI?

- If you want to do stuff to the game's UI then take a look at the [HUD native namespace](https://nativedb.dotindustries.dev/gta5/natives/0xABA17D7CE615ADBF)

- If you meant YimMenu's GUI (your script's UI) then read the [ImGui docs](https://github.com/YimMenu/YimMenu/blob/master/docs/lua/tables/ImGui.md)

Best of luck bud ♥

**Reply by MindOfBog** [reply]
I meant the games UI, but not really like the UI elements, more so unlocking locked menu options and stuff. Thank you for your help o7
